### PR TITLE
Messaging - Manager layer

### PR DESCRIPTION
**Pull request by kaysoky**
- Extra table to link Nodes to Messages
- Extra DAO-level methods for fetching threads without filtering
- Change Message DTO to have a unique set of recipients
- Manager including permission checking
- Method to-be-used by the message sending worker

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMessageDAOImpl.java [hunk 1]
```diff
@@ -239,4 +278,35 @@ public void updateMessageStatus(String messageId, String userId,
...
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	public void registerThreadToNode(String threadId, String nodeId) {
+		DBONodeMessages dbo = new DBONodeMessages();
+		dbo.setNodeId(KeyFactory.stringToKey(nodeId));
+		dbo.setThreadId(Long.parseLong(threadId));
+		basicDAO.createNew(dbo);
+	}
+
+	@Override
+	public String getThreadOfNode(String nodeId) throws NotFoundException {
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMessageDAOImpl.java]: next revision will support multiple threads on a node

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMessageDAOImpl.java [hunk 2]
```diff
@@ -239,4 +278,35 @@ public void updateMessageStatus(String messageId, String userId,
...
+			throw new NotFoundException("There is no entity linked to the provided node (" + nodeId + ")");
+		}
+	}
+
+	@Override
+	public String getNodeOfThread(String threadId) throws NotFoundException {
+		MapSqlParameterSource params = new MapSqlParameterSource();
+		params.addValue(THREAD_ID_PARAM_NAME, threadId);
+		try {
+			return simpleJdbcTemplate.queryForObject(SELECT_NODE_ID_FROM_THREAD_ID, String.class, params);
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOMessageDAOImpl.java]: use key factory to add 'syn' prefix

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBONode.java [hunk 3]
```diff
@@ -205,10 +205,12 @@ public DBONode createBackupFromDatabaseObject(DBONode dbo) {
 		return DBONode.class;
 	}
 	
+	@SuppressWarnings("rawtypes")
 	@Override
 	public List<MigratableDatabaseObject> getSecondaryTypes() {
 		List<MigratableDatabaseObject> list = new LinkedList<MigratableDatabaseObject>();
 		list.add(new DBORevision());
+		list.add(new DBONodeMessages());
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBONode.java]: DBONodeMesageThread?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/query/jdo/SqlConstants.java [hunk 4]
```diff
@@ -159,6 +159,12 @@
  	public static final String COL_MESSAGE_STATUS_RECIPIENT_ID      = "RECIPIENT_ID";
  	public static final String COL_MESSAGE_STATUS                   = "STATUS";
  	public static final String DDL_MESSAGE_STATUS                   = "schema/MessageStatus-ddl.sql";
+ 	
+ 	// Node message
+ 	public static final String TABLE_NODE_MESSAGES                  = "NODE_MESSAGES";
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/query/jdo/SqlConstants.java]: ..._NODE_MESSAGE_THREAD

File: lib/jdomodels/src/main/resources/schema/NodeMessages-ddl.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE `NODE_MESSAGES` (
+  `NODE_ID` bigint(20) NOT NULL,
+  `THREAD_ID` bigint(20) NOT NULL,
+  PRIMARY KEY (`NODE_ID`),
```
Comment (brucehoff) [inline lib/jdomodels/src/main/resources/schema/NodeMessages-ddl.sql]: primary key is node + thread

File: lib/jdomodels/src/main/resources/schema/NodeMessages-ddl.sql [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  UNIQUE KEY (`THREAD_ID`),
+  CONSTRAINT `NODE_ID_FK` FOREIGN KEY (`NODE_ID`) REFERENCES `JDONODE` (`ID`) ON DELETE CASCADE,
+  CONSTRAINT `THREAD_ID_FK` FOREIGN KEY (`THREAD_ID`) REFERENCES `MESSAGE` (`THREAD_ID`) ON DELETE CASCADE
+)
```
Comment (brucehoff) [inline lib/jdomodels/src/main/resources/schema/NodeMessages-ddl.sql]: make sure that deletion of a message does NOT delete the join table entry.  Add a DAO level test which creates two messages with one thread, deletes one, then queries on the node/thread relationship.

**Comment by kaysoky** [comment]
Needs major DAO level changes first...

**Pull request closed without merging**